docs/Compiler_Hardening_Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md

```diff
@@ -213,22 +212,24 @@ For example, developers can decide to promote warnings that indicate interferenc
 | <span id="-D_FORTIFY_SOURCE=2">`-D_FORTIFY_SOURCE=2`</span><br/>(requires `-O1` or higher) | GCC 4.0<br/>Clang 5.0.0[^1] | In addition to checks covered by `-D_FORTIFY_SOURCE=1`, also trap code that may be conforming to the C standard but still unsafe |
+| <span id="-D_FORTIFY_SOURCE=3">`-D_FORTIFY_SOURCE=3`</span><br/>(requires `-O1` or higher) | GCC 12.0<br/>Clang 9.0.0[^1] | Same checks as in `-D_FORTIFY_SOURCE=2`, but with significantly more calls fortified with a potential to impact performance in some rare cases |
 
 #### Synopsis
 
-The `_FORTIFY_SOURCE` macro enables a set of extension to the GNU C library (glibc) that will make the compiler automatically add buffer overflow checks to calls to the following functions:  
+The `_FORTIFY_SOURCE` macro enables a set of extensions to the GNU C library (glibc) that enable checking at entry points of a number of functions to immediately abort execution when it encounters unsafe behavior. A key feature of this checking is validation of objects passed to these function calls to ensure that the call will not result in a buffer overflow. This relies on the compiler being able to compute the size of the protected object at compile time. A full list of these functions is maintained in the GNU C Library manual[^28]:
 
 > memcpy, mempcpy, memmove, memset, strcpy, stpcpy, strncpy, strcat, strncat, sprintf, vsprintf, snprintf, vsnprintf, gets
 
-The `_FORTIFY_SOURCE` mechanisms have two modes of operation:
+The `_FORTIFY_SOURCE` mechanisms have three modes of operation:
 
-- `-D_FORTIFY_SOURCE=1`: conservative, compile-time and runtime checks; will not change (defined) behavior of programs
+- `-D_FORTIFY_SOURCE=1`: conservative, compile-time and runtime checks; will not change (defined) behavior of programs. Checking for overflows is enabled when the compiler is able to estimate a compile time constant size for the protected object.
 - `-D_FORTIFY_SOURCE=2`: stricter checks that also detect behavior that may be unsafe even though it conforms to the C standard; may affect program behavior by disallowing certain programming constructs. An example of such checks is restricting of the `%n` format specifier to read-only format strings.
+- `-D_FORTIFY_SOURCE=3`: Same checks as those covered by `-D_FORTIFY_SOURCE=3` except that checking is enabled even when the compiler is able to estimate the size of the protected object as an expression, not just a compile time constant.
 
 To benefit from `_FORTIFY_SOURCE` checks following requirements must be met:  
 
-- the application must be built with -O1 optimizations or higher
-- the sizes of the destination buffers must be known at compile time, and
+- the application must be built with `-O1` optimizations or higher; at least `-O2` is recommended.
```

Comment:
Is there a benefit to `_FORTIFY_SOURCE` behavior between `-O1` and `-O2´ or is the recommendation simply motivated by better overall performance?

Reply:
`_FORTIFY_SOURCE` depends on object size estimation and that improves with optimization, so `_FORTIFY_SOURCE` coverage improves with better optimization.